File: desktop/news/aggregator.py
"""Merges raw items from every source into enriched, deduped, sorted NewsItem[].
Mirrors src/lib/news/aggregator.ts."""
from __future__ import annotations

import re
from datetime import datetime, timedelta, timezone

from .analysis import score_impact, score_importance, score_sentiment, summarize
from .categorize import categorize
from .types import NewsCategory, NewsItem, SourceResult


def _normalize_title(headline: str) -> str:
    lowered = re.sub(r"[^a-z0-9 ]", "", headline.lower())
    return re.sub(r"\s+", " ", lowered).strip()


def _parse_time(iso: str) -> datetime:
    d = datetime.fromisoformat(iso.replace("Z", "+00:00"))
    return d if d.tzinfo else d.replace(tzinfo=timezone.utc)


def _enrich(raw) -> NewsItem:
    text = f"{raw.headline} {raw.summary or ''}"
    categories, sector = categorize(text)
    if raw.provided_sentiment:
        sentiment, confidence = raw.provided_sentiment.label, raw.provided_sentiment.score
    else:
        sentiment, confidence = score_sentiment(text)
    importance = score_importance(text, categories, confidence)
    return NewsItem(
        id=raw.id,
        headline=raw.headline,
        source=raw.source,
        time=raw.time,
        categories=categories,
        sector=sector,
        ai_summary=summarize(raw.headline, raw.summary),
        sentiment=sentiment,
        sentiment_confidence=confidence,
        importance=importance,
        impact=score_impact(importance, confidence),
        summary=raw.summary,
        url=raw.url,
        publisher=raw.publisher,
        tickers=list(raw.tickers or []),
    )


def aggregate(results: list[SourceResult]) -> list[NewsItem]:
    """Combines source results into a single enriched, deduped, newest-first feed."""
    seen: dict[str, NewsItem] = {}
    for result in results:
        for raw in result.items:
            key = raw.url or _normalize_title(raw.headline)
            existing = seen.get(key)
            if existing:
                existing.tickers = list(dict.fromkeys([*existing.tickers, *(raw.tickers or [])]))
                continue
            seen[key] = _enrich(raw)
    return sorted(seen.values(), key=lambda i: _parse_time(i.time), reverse=True)


def within_hours(items: list[NewsItem], hours: float) -> list[NewsItem]:
    cutoff = datetime.now(timezone.utc) - timedelta(hours=hours)
    return [i for i in items if _parse_time(i.time) >= cutoff]


def within_days(items: list[NewsItem], days: float) -> list[NewsItem]:
    return within_hours(items, days * 24)


def group_by_category(items: list[NewsItem]) -> dict[NewsCategory, list[NewsItem]]:
    """Groups items by every category they matched (a story can land in >1 bucket)."""
    grouped: dict[NewsCategory, list[NewsItem]] = {}
    for item in items:
        for cat in item.categories:
            grouped.setdefault(cat, []).append(item)
    return grouped


def top_by_importance(items: list[NewsItem], n: int) -> list[NewsItem]:
    return sorted(items, key=lambda i: i.importance, reverse=True)[:n]

File: src/lib/news/aggregator.ts
/** Merges raw items from every source into enriched, deduped, sorted NewsItem[]. */
import { categorize } from './categorize';
import { scoreImpact, scoreImportance, scoreSentiment, summarize } from './analysis';
import type { NewsCategory, NewsItem, RawNewsItem } from './types';
import type { SourceResult } from './sources';

function normalizeTitle(headline: string): string {
  return headline
    .toLowerCase()
    .replace(/[^a-z0-9 ]/g, '')
    .replace(/\s+/g, ' ')
    .trim();
}

function enrich(raw: RawNewsItem): NewsItem {
  const text = `${raw.headline} ${raw.summary ?? ''}`;
  const { categories, sector } = categorize(text);
  const sentiment = raw.providedSentiment
    ? { label: raw.providedSentiment.label, confidence: raw.providedSentiment.score }
    : scoreSentiment(text);
  const importance = scoreImportance(text, categories, sentiment.confidence);
  return {
    ...raw,
    categories,
    sector,
    aiSummary: summarize(raw.headline, raw.summary),
    sentiment: sentiment.label,
    sentimentConfidence: sentiment.confidence,
    importance,
    impact: scoreImpact(importance, sentiment.confidence),
  };
}

/** Combines source results into a single enriched, deduped, newest-first feed. */
export function aggregate(results: SourceResult[]): NewsItem[] {
  const seen = new Map<string, NewsItem>();
  for (const result of results) {
    for (const raw of result.items) {
      const key = raw.url || normalizeTitle(raw.headline);
      const existing = seen.get(key);
      if (existing) {
        // Merge tickers from the duplicate hit (same story from two sources).
        existing.tickers = Array.from(new Set([...(existing.tickers ?? []), ...(raw.tickers ?? [])]));
        continue;
      }
      seen.set(key, enrich(raw));
    }
  }
  return Array.from(seen.values()).sort((a, b) => +new Date(b.time) - +new Date(a.time));
}

export function withinHours(items: NewsItem[], hours: number): NewsItem[] {
  const cutoff = Date.now() - hours * 3_600_000;
  return items.filter((i) => +new Date(i.time) >= cutoff);
}

export function withinDays(items: NewsItem[], days: number): NewsItem[] {
  return withinHours(items, days * 24);
}

/** Groups items by every category they matched (a story can land in >1 bucket). */
export function groupByCategory(items: NewsItem[]): Map<NewsCategory, NewsItem[]> {
  const map = new Map<NewsCategory, NewsItem[]>();
  for (const item of items) {
    for (const cat of item.categories) {
      const bucket = map.get(cat) ?? [];
      bucket.push(item);
      map.set(cat, bucket);
    }
  }
  return map;
}

export function topByImportance(items: NewsItem[], n: number): NewsItem[] {
  return [...items].sort((a, b) => b.importance - a.importance).slice(0, n);
}
